Subject: [PATCH] buck 2
buck
fixup
Add roaring bitmap array
---
Index: CMake/resolve_dependency_modules/roaring.cmake
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/CMake/resolve_dependency_modules/roaring.cmake b/CMake/resolve_dependency_modules/roaring.cmake
new file mode 100644
--- /dev/null	(revision 7727c6a00aaa245c97d8c897e707fbbf4e8b7d82)
+++ b/CMake/resolve_dependency_modules/roaring.cmake	(revision 7727c6a00aaa245c97d8c897e707fbbf4e8b7d82)
@@ -0,0 +1,35 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+include_guard(GLOBAL)
+
+set(VELOX_ROARING_VERSION 2.0.4)
+set(
+  VELOX_ROARING_BUILD_SHA256_CHECKSUM
+  3c962c196ba28abf2639067f2e2fd25879744ba98152a4e0e74556ca515eda33
+)
+set(
+  VELOX_ROARING_SOURCE_URL
+  "https://github.com/RoaringBitmap/CRoaring/archive/refs/tags/v${VELOX_ROARING_VERSION}.tar.gz"
+)
+
+velox_resolve_dependency_url(ROARING)
+
+FetchContent_Declare(
+  roaring
+  URL ${VELOX_ROARING_SOURCE_URL}
+  URL_HASH ${VELOX_ROARING_BUILD_SHA256_CHECKSUM}
+)
+
+FetchContent_MakeAvailable(roaring)
Index: CMakeLists.txt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/CMakeLists.txt b/CMakeLists.txt
--- a/CMakeLists.txt	(revision 2593702d65456699919f8d32ad98fbf887a793ef)
+++ b/CMakeLists.txt	(revision 7727c6a00aaa245c97d8c897e707fbbf4e8b7d82)
@@ -717,4 +717,7 @@
   velox_resolve_dependency(geos)
 endif()

+velox_set_source(roaring)
+velox_resolve_dependency(roaring)
+
 add_subdirectory(velox)
Index: velox/functions/sparksql/CMakeLists.txt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/velox/functions/sparksql/CMakeLists.txt b/velox/functions/sparksql/CMakeLists.txt
--- a/velox/functions/sparksql/CMakeLists.txt	(revision 2593702d65456699919f8d32ad98fbf887a793ef)
+++ b/velox/functions/sparksql/CMakeLists.txt	(revision 7727c6a00aaa245c97d8c897e707fbbf4e8b7d82)
@@ -29,6 +29,7 @@
   MakeTimestamp.cpp
   Map.cpp
   RegexFunctions.cpp
+  RoaringBitmapArray.cpp
   Size.cpp
   String.cpp
   UnscaledValueFunction.cpp
@@ -42,6 +43,7 @@
   velox_functions_util
   Folly::folly
   simdjson::simdjson
+  roaring
 )

 if(NOT VELOX_MONO_LIBRARY)
Index: velox/functions/sparksql/RoaringBitmapArray.cpp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/velox/functions/sparksql/RoaringBitmapArray.cpp b/velox/functions/sparksql/RoaringBitmapArray.cpp
new file mode 100644
--- /dev/null	(revision fd4489e96f78465546b6e0eedcfd475a4d4912d7)
+++ b/velox/functions/sparksql/RoaringBitmapArray.cpp	(revision fd4489e96f78465546b6e0eedcfd475a4d4912d7)
@@ -0,0 +1,178 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/functions/sparksql/RoaringBitmapArray.h"
+#include "common/base/IOUtils.h"
+
+namespace facebook::velox::functions::sparksql {
+
+namespace {
+const int64_t kMaxRepresentableValue =
+    RoaringBitmapArray::composeFromHighLowBytes(
+        std::numeric_limits<int32_t>::max() - 1,
+        std::numeric_limits<int32_t>::max());
+}
+
+void RoaringBitmapArray::deserialize(const char* serialized) {
+  bitmaps_.clear();
+  buckContexts_.clear();
+  common::InputByteStream stream(serialized);
+  const auto magicNumber = stream.read<int32_t>();
+  switch (magicNumber) {
+    case kPortableSerializationFormatMagicNumber: {
+      const auto numberOfBitmaps = stream.read<int64_t>();
+      VELOX_CHECK_GE(
+          numberOfBitmaps,
+          0,
+          "Invalid RoaringBitmapArray length ({} < 0)",
+          numberOfBitmaps);
+      VELOX_CHECK_LE(
+          numberOfBitmaps,
+          std::numeric_limits<int32_t>::max(),
+          "Invalid RoaringBitmapArray length ({} > {})",
+          numberOfBitmaps,
+          std::numeric_limits<int32_t>::max());
+      // The sparse bitmap array format uses the number as an array size lower
+      // bound.
+      const auto minimumArraySize = static_cast<int32_t>(numberOfBitmaps);
+      bitmaps_.reserve(minimumArraySize);
+      buckContexts_.reserve(minimumArraySize);
+      int32_t lastIndex = 0;
+      for (int32_t i = 0; i < numberOfBitmaps; ++i) {
+        const auto key = stream.read<int32_t>();
+        VELOX_CHECK_GE(
+            key, 0, "Invalid unsigned entry in RoaringBitmapArray ({})", key);
+        VELOX_CHECK_GE(
+            key,
+            lastIndex,
+            "Keys are required to be sorted in ascending order");
+        // Fill gaps in sparse data.
+        while (lastIndex < key) {
+          bitmaps_.emplace_back(std::make_shared<roaring::Roaring>());
+          buckContexts_.emplace_back(std::make_shared<roaring::BulkContext>());
+          ++lastIndex;
+        }
+        roaring::api::roaring_bitmap_t* r =
+            roaring::api::roaring_bitmap_portable_deserialize(
+                serialized + stream.offset());
+        VELOX_CHECK_NOT_NULL(r);
+        bitmaps_.emplace_back(std::make_shared<roaring::Roaring>(r));
+        buckContexts_.emplace_back(std::make_shared<roaring::BulkContext>());
+        ++lastIndex;
+        // Advances the stream for N bytes which is the serialized size of the
+        // previous read bitmap.
+        auto sizeToAdvance = bitmaps_.back()->getSizeInBytes();
+        for (size_t j = 0; j < sizeToAdvance; j++) {
+          stream.read<int8_t>();
+        }
+      }
+    } break;
+    default:
+      VELOX_NYI(
+          "Unexpected magic number in serialized roaring bitmap array: {}",
+          magicNumber);
+  }
+}
+
+void RoaringBitmapArray::serialize(char* buf) {
+  common::OutputByteStream stream(buf);
+  stream.appendOne<int32_t>(kPortableSerializationFormatMagicNumber);
+  stream.appendOne<int64_t>(bitmaps_.size());
+  for (size_t i = 0; i < bitmaps_.size(); ++i) {
+    stream.appendOne<int32_t>(i);
+    const auto serializedSize = bitmaps_[i]->getSizeInBytes();
+    // FIXME: The following code conducts an extra copy.
+    std::vector<char> data(serializedSize);
+    bitmaps_[i]->write(data.data(), data.size());
+    stream.append(data.data(), data.size());
+  }
+}
+
+void RoaringBitmapArray::add(int64_t value) {
+  checkValue(value);
+  const auto high = highBytes(value);
+  const auto low = lowBytes(value);
+  if (high >= bitmaps_.size()) {
+    // Grows the bitmap array.
+    for (int32_t i = bitmaps_.size(); i <= high; ++i) {
+      bitmaps_.emplace_back(std::make_shared<roaring::Roaring>());
+      buckContexts_.emplace_back(std::make_shared<roaring::BulkContext>());
+    }
+  }
+  auto bitmap = bitmaps_[high];
+  bitmap->add(low);
+}
+
+bool RoaringBitmapArray::contains(int64_t value) {
+  checkValue(value);
+  const auto high = highBytes(value);
+  if (high >= bitmaps_.size()) {
+    return false;
+  }
+  if (FOLLY_LIKELY(high == lastHighBytes_)) {
+    // Fast path for ordered input.
+    const auto low = lowBytes(value);
+    return lastBitmap_->containsBulk(*lastContext_, low);
+  }
+  const auto highBitmap = bitmaps_[high];
+  const auto highBuckContext = buckContexts_[high];
+  const auto low = lowBytes(value);
+  lastHighBytes_ = high;
+  lastBitmap_ = highBitmap.get();
+  lastContext_ = highBuckContext.get();
+  return highBitmap->containsBulk(*highBuckContext, low);
+}
+
+int64_t RoaringBitmapArray::serializedSizeInBytes() {
+  const int64_t magicNumberSize = 4L;
+  const int64_t bitmapCountSize = 8L;
+  const int64_t individualBitmapKeySize = 4L;
+
+  int64_t sum = 0L;
+  sum += magicNumberSize;
+  sum += bitmapCountSize;
+
+  for (const auto& bitmap : bitmaps_) {
+    sum += individualBitmapKeySize;
+    sum += bitmap->getSizeInBytes();
+  }
+
+  return sum;
+}
+
+void RoaringBitmapArray::checkValue(int64_t value) {
+  VELOX_CHECK_GE(value, 0, "Invalid RoaringBitmapArray value ({}) >= 0", value);
+  VELOX_CHECK_LE(
+      value,
+      kMaxRepresentableValue,
+      "Invalid RoaringBitmapArray value ({}) <= {}",
+      value,
+      kMaxRepresentableValue);
+}
+
+int32_t RoaringBitmapArray::highBytes(int64_t value) {
+  return static_cast<int32_t>(value >> 32);
+}
+
+int32_t RoaringBitmapArray::lowBytes(int64_t value) {
+  return static_cast<int32_t>(value & 0xFFFFFFFFLL);
+}
+
+int64_t RoaringBitmapArray::composeFromHighLowBytes(int32_t high, int32_t low) {
+  return static_cast<int64_t>(high) << 32 |
+      (static_cast<int64_t>(low) & 0xFFFFFFFFLL);
+}
+} // namespace facebook::velox::functions::sparksql
Index: velox/functions/sparksql/RoaringBitmapArray.h
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/velox/functions/sparksql/RoaringBitmapArray.h b/velox/functions/sparksql/RoaringBitmapArray.h
new file mode 100644
--- /dev/null	(revision fd4489e96f78465546b6e0eedcfd475a4d4912d7)
+++ b/velox/functions/sparksql/RoaringBitmapArray.h	(revision fd4489e96f78465546b6e0eedcfd475a4d4912d7)
@@ -0,0 +1,83 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "common/base/Exceptions.h"
+
+#include <core/QueryConfig.h>
+#include <functions/Macros.h>
+#include <roaring.hh>
+
+namespace facebook::velox::functions::sparksql {
+
+class RoaringBitmapArray {
+ public:
+  RoaringBitmapArray() : bitmaps_({}) {}
+
+  void deserialize(const char* serialized);
+
+  void serialize(char* buf);
+
+  void add(int64_t value);
+
+  bool contains(int64_t value);
+
+  int64_t serializedSizeInBytes();
+
+  // Composes a pair (high, low).
+  static int64_t composeFromHighLowBytes(int32_t high, int32_t low);
+
+ private:
+  static void checkValue(int64_t value);
+
+  // Extracts the high 32 bits.
+  static int32_t highBytes(int64_t value);
+
+  // Extracts the low 32 bits.
+  static int32_t lowBytes(int64_t value);
+
+  static constexpr int32_t kPortableSerializationFormatMagicNumber = 1681511377;
+  std::vector<std::shared_ptr<roaring::Roaring>> bitmaps_;
+  std::vector<std::shared_ptr<roaring::BulkContext>> buckContexts_;
+  int32_t lastHighBytes_{-1};
+  roaring::Roaring* lastBitmap_{nullptr};
+  roaring::BulkContext* lastContext_{nullptr};
+};
+
+template <typename T>
+struct RoaringBitmapArrayContains {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  void initialize(
+      const std::vector<TypePtr>& /*inputTypes*/,
+      const core::QueryConfig&,
+      const arg_type<Varbinary>* serialized,
+      const arg_type<int64_t>*) {
+    if (serialized != nullptr) {
+      array_.deserialize(serialized->str().c_str());
+    }
+  }
+
+  FOLLY_ALWAYS_INLINE void
+  call(bool& result, const arg_type<Varbinary>&, const int64_t& input) {
+    result = array_.contains(input);
+  }
+
+ private:
+  RoaringBitmapArray array_{};
+};
+} // namespace facebook::velox::functions::sparksql
Index: velox/functions/sparksql/registration/RegisterBinary.cpp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/velox/functions/sparksql/registration/RegisterBinary.cpp b/velox/functions/sparksql/registration/RegisterBinary.cpp
--- a/velox/functions/sparksql/registration/RegisterBinary.cpp	(revision 2593702d65456699919f8d32ad98fbf887a793ef)
+++ b/velox/functions/sparksql/registration/RegisterBinary.cpp	(revision 7727c6a00aaa245c97d8c897e707fbbf4e8b7d82)
@@ -17,6 +17,7 @@
 #include "velox/functions/prestosql/BinaryFunctions.h"
 #include "velox/functions/sparksql/Hash.h"
 #include "velox/functions/sparksql/MightContain.h"
+#include "velox/functions/sparksql/RoaringBitmapArray.h"
 #include "velox/functions/sparksql/String.h"
 
 namespace facebook::velox::functions::sparksql {
@@ -44,6 +45,8 @@
       {prefix + "sha1"});
   registerFunction<Sha2HexStringFunction, Varchar, Varbinary, int32_t>(
       {prefix + "sha2"});
+  registerFunction<RoaringBitmapArrayContains, bool, Varbinary, int64_t>(
+      {prefix + "roaring_bitmap_array_contains"});
 }
 
 } // namespace facebook::velox::functions::sparksql
Index: velox/functions/sparksql/tests/CMakeLists.txt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/velox/functions/sparksql/tests/CMakeLists.txt b/velox/functions/sparksql/tests/CMakeLists.txt
--- a/velox/functions/sparksql/tests/CMakeLists.txt	(revision 2593702d65456699919f8d32ad98fbf887a793ef)
+++ b/velox/functions/sparksql/tests/CMakeLists.txt	(revision 7727c6a00aaa245c97d8c897e707fbbf4e8b7d82)
@@ -63,6 +63,7 @@
   RandTest.cpp
   ReadSidePaddingFunctionTest.cpp
   RegexFunctionsTest.cpp
+  RoaringBitmapArrayTest.cpp
   SliceTest.cpp
   SizeTest.cpp
   SortArrayTest.cpp
Index: velox/functions/sparksql/tests/RoaringBitmapArrayTest.cpp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/velox/functions/sparksql/tests/RoaringBitmapArrayTest.cpp b/velox/functions/sparksql/tests/RoaringBitmapArrayTest.cpp
new file mode 100644
--- /dev/null	(revision 7727c6a00aaa245c97d8c897e707fbbf4e8b7d82)
+++ b/velox/functions/sparksql/tests/RoaringBitmapArrayTest.cpp	(revision 7727c6a00aaa245c97d8c897e707fbbf4e8b7d82)
@@ -0,0 +1,55 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <functions/sparksql/RoaringBitmapArray.h>
+
+#include "velox/core/Expressions.h"
+#include "velox/functions/sparksql/tests/SparkFunctionBaseTest.h"
+
+namespace facebook::velox::functions::sparksql::test {
+namespace {
+
+class RoaringBitmapArrayTest : public ::testing::Test {};
+
+TEST_F(RoaringBitmapArrayTest, contains) {
+  RoaringBitmapArray array{};
+  array.add(206L);
+  array.add(10L << 32 | 10L);
+  EXPECT_TRUE(array.contains(206L));
+  EXPECT_FALSE(array.contains(207L));
+  EXPECT_TRUE(array.contains(10L << 32 | 10));
+  EXPECT_FALSE(array.contains(11L << 32 | 10));
+  EXPECT_FALSE(array.contains(10L << 32 | 11));
+}
+
+TEST_F(RoaringBitmapArrayTest, serde) {
+  RoaringBitmapArray array{};
+  array.add(206L);
+  array.add(10L << 32 | 10L);
+  std::string data;
+  data.resize(array.serializedSizeInBytes());
+  array.serialize(data.data());
+  RoaringBitmapArray deserialized{};
+  deserialized.deserialize(data.data());
+  EXPECT_TRUE(deserialized.contains(206L));
+  EXPECT_FALSE(deserialized.contains(207L));
+  EXPECT_TRUE(deserialized.contains(10L << 32 | 10));
+  EXPECT_FALSE(deserialized.contains(11L << 32 | 10));
+  EXPECT_FALSE(deserialized.contains(10L << 32 | 11));
+}
+
+} // namespace
+} // namespace facebook::velox::functions::sparksql::test
