Index: velox/vector/VariantToVector.cpp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/velox/vector/VariantToVector.cpp b/velox/vector/VariantToVector.cpp
--- a/velox/vector/VariantToVector.cpp	(revision 1d12342246b75ccdfd2ac050a0adf27164a8f893)
+++ b/velox/vector/VariantToVector.cpp	(revision 9a23bdbc582e65651009bcd6c093c81ea7191ddb)
@@ -62,7 +62,14 @@
 
     // Allocate nulls and data buffers and set all values to null by default.
     const vector_size_t dataSize = data.size();
-    BufferPtr valuesBuffer = AlignedBuffer::allocate<T>(dataSize, pool);
+    BufferPtr valuesBuffer;
+    if constexpr (std::is_same_v<T, StringView>) {
+      // Make sure to initialize StringView values so they can be safely
+      // accessed.
+      valuesBuffer = AlignedBuffer::allocate<T>(dataSize, pool, T());
+    } else {
+      valuesBuffer = AlignedBuffer::allocate<T>(dataSize, pool);
+    }
     BufferPtr nulls = allocateNulls(dataSize, pool, bits::kNull);
 
     // Create flat vector to store all the values.
